CoreText font platform should respect legacy visionOS styles when relevant

hortont424 opened the pull request:
#### eb8b191e357fdeb6d91cfc181bbffd10c3ea210d
<pre>
CoreText font platform should respect legacy visionOS styles when relevant
<a href="https://bugs.webkit.org/show_bug.cgi?id=260042">https://bugs.webkit.org/show_bug.cgi?id=260042</a>
rdar://113417307

Reviewed by Aditya Keerthi and Mike Wyrzykowski.

The Web Content process needs to follow the host application&apos;s sub-idiom
for the purposes of font lookup, instead of just getting the default table
for its bundle identifier, to ensure correct resolution of system font styles.

* Source/WebCore/Configurations/WebCore.xcconfig:
Link RuntimeSupport on visionOS.

* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.h:
* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.mm:
(PAL::currentUserInterfaceIdiomIsVisionOrVisionLegacy):
(PAL::currentUserInterfaceIdiomIsVisionLegacy):
Add getters that can determine which visionOS sub-idiom you&apos;re using.
All clients (the vast majority) that don&apos;t care about the points-per-meter
value should use `IsVisionOrVisionLegacy`.

(PAL::determineVisionSubidiom):
Determine the sub-idiom by checking _RSCurrentProcessUsesNewPointsPerMeter.

(PAL::updateCurrentUserInterfaceIdiom):
(PAL::currentUserInterfaceIdiomIsVision): Deleted.
* Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.cpp:
(WebCore::fontPlatform):
Determine the correct CoreText font platform based on the sub-idiom.

* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultAlternateFormControlDesignEnabled):
(WebKit::defaultVideoFullscreenRequiresElementFullscreen):
Adopt the new getters.

Canonical link: <a href="https://commits.webkit.org/266799@main">https://commits.webkit.org/266799@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2b92b5aba61efb1f70720044153efe5ecf67fbb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14788 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15093 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16535 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/13927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14925 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17614 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15193 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14968 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15460 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12560 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17269 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12743 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13343 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20322 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13818 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13510 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16743 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14054 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11882 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13345 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3570 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17678 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13897 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.mm:
```diff
@@ -44,11 +48,19 @@ bool currentUserInterfaceIdiomIsSmallScreen()
+{
+    if (!s_currentUserInterfaceIdiom.load())
+        updateCurrentUserInterfaceIdiom();
+    auto idiom = *s_currentUserInterfaceIdiom.load();
+    return idiom == UserInterfaceIdiom::Vision || idiom == UserInterfaceIdiom::VisionLegacy;
```
Optional nit: you don't need to unwrap the value because of this `operator==` overload.

```cpp
template< class T, class U >
constexpr bool operator==( const optional<T>& opt, const U& value );
```

mwyrzykowski replied:
but then wouldn't we be performing two atomic loads?

hortont424 replied:
no, I think it's still just one load (we'd load into a local optional<uint...>). But I don't love the thought of it :P

aprotyas reviewed (commented):
(lgtm)

hortont424 committed:
CoreText font platform should respect legacy visionOS styles when relevant
https://bugs.webkit.org/show_bug.cgi?id=260042
rdar://113417307

Reviewed by Aditya Keerthi and Mike Wyrzykowski.

The Web Content process needs to follow the host application's sub-idiom
for the purposes of font lookup, instead of just getting the default table
for its bundle identifier, to ensure correct resolution of system font styles.

* Source/WebCore/Configurations/WebCore.xcconfig:
Link RuntimeSupport on visionOS.

* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.h:
* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.mm:
(PAL::currentUserInterfaceIdiomIsVisionOrVisionLegacy):
(PAL::currentUserInterfaceIdiomIsVisionLegacy):
Add getters that can determine which visionOS sub-idiom you're using.
All clients (the vast majority) that don't care about the points-per-meter
value should use `IsVisionOrVisionLegacy`.

(PAL::determineVisionSubidiom):
Determine the sub-idiom by checking _RSCurrentProcessUsesNewPointsPerMeter.

(PAL::updateCurrentUserInterfaceIdiom):
(PAL::currentUserInterfaceIdiomIsVision): Deleted.
* Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.cpp:
(WebCore::fontPlatform):
Determine the correct CoreText font platform based on the sub-idiom.

* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultAlternateFormControlDesignEnabled):
(WebKit::defaultVideoFullscreenRequiresElementFullscreen):
Adopt the new getters.

Canonical link: https://commits.webkit.org/266799@main
